### PR TITLE
[on hold] Use cmake

### PR DESCRIPTION
**Pull request by ltalirz**
Try using the new cmake approach (`configure` is proving difficult...)

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ltalirz** [comment]
Wait until https://gitlab.com/QEF/q-e/-/issues/298 is fixed

**Comment by ltalirz** [comment]
Fixed now; revisit this PR with the next QE release 
https://gitlab.com/QEF/q-e/-/merge_requests/1423

**Comment by ltalirz** [comment]
@conda-forge-admin, please rerender

**Comment by ltalirz** [comment]
@ye-luo I'm getting [an error](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=354265&view=logs&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d&l=1798) 
```
make[3]: *** No rule to make target `qe_plotphon_eminmax_exe', needed by `CMakeFiles/ph'.  Stop.
```
with the install line
```
make pwall ld1 cp epw
```

Is this possibly a bug or am I doing something wrong here?

**Comment by ye-luo** [comment]
@ltalirz this is an issue spotted after 6.8 release being created. https://gitlab.com/QEF/q-e/-/wikis/Support/CMake-glitches. It has been fixed in develop and CI has just been added.

File: recipe/build.sh [hunk 1]
```diff
@@ -1,46 +1,31 @@
...
+    -DQE_ENABLE_TEST=ON \
+    -DCMAKE_INSTALL_PREFIX=${PREFIX}
+    
+make pwall ld1 cp epw
+
+## see https://gitlab.com/QEF/q-e/-/blob/develop/test-suite/Makefile
+#cd test-suite
+#make run-travis
+#cd ..
+make test
```
Comment (ye-luo) [inline recipe/build.sh:29]: make test triggers all the tests but only pw, cp and unit tests are safe to run when you use CMake.
So better testing `ctest -L "pw|cp|unit" -LE epw` for the moment.

**Comment by ltalirz** [comment]
Thanks for the explanation!

> With 6.8, you may just `make` without any targets and this error won't show up.

I started with this but the previous version of the QE conda package included executables like `phcg.x` which was not created by simply running `make`.
Which commit fixed the cmake list (so that I can apply it here as a patch)?

File: recipe/build.sh (hunk 1 above)
Comment (ltalirz) [reply recipe/build.sh:29]: thanks, will do!

**Comment by ye-luo** [comment]
> Thanks for the explanation!
> 
> > With 6.8, you may just `make` without any targets and this error won't show up.
> 
> I started with this but the previous version of the QE conda package included executables like `phcg.x` which was not created by simply running `make`.

This is an unsolvable issue of the manually written makefiles. In CMake, `make` builds all the targets reachable by CMake. It seems that phcg.x is not connected in CMake, will fix that soon.

> Which commit fixed the cmake list (so that I can apply it here as a patch)?

d3b64c6b

**Comment by ltalirz** [comment]
> In CMake, `make` builds all the targets reachable by CMake.

Ah ok, I didn't realize that.



> It seems that phcg.x is not connected in CMake, will fix that soon.

Cheers, let me know!

In the meanwhile I'll remove the check for phcg.x and see whether that fixes things.

**Comment by ye-luo** [comment]
phcg.x has been added in develop https://gitlab.com/QEF/q-e/-/merge_requests/1516

**Comment by ltalirz** [comment]
Thanks @ye-luo !

I have two more small questions:

 * a couple of tests mention the need for libxc. I've added the library to the build environment but it seems it was not automatically recognized. Is there an option to enable it (I searched for libxc in the top-level cmake list and didn't find anything)?
 * for this openmpi build, on conda-forge I need to run any parallel tests with `mpiexec --bind-to none -mca plm isolated`
   Is there a way to pass along command line options to how `ctest` uses mpi?

**Comment by ye-luo** [comment]
phcg.x is now connected but needs two commits https://gitlab.com/QEF/q-e/-/commit/5185853e763c1bfb2ba33e01f19097c0bb400cc1 and https://gitlab.com/QEF/q-e/-/commit/c5373212f2717f916e6ee200aad4197b8e43a901

>     * a couple of tests mention the need for libxc. I've added the library to the build environment but it seems it was not automatically recognized. Is there an option to enable it (I searched for libxc in the top-level cmake list and didn't find anything)?

-DQE_ENABLE_LIBXC=ON should be at the top level CMakeLists.txt.If the automatic search failed, add -DCMAKE_PREFIX_PATH=your_libxc.


>     * for this openmpi build, on conda-forge I need to run any parallel tests with `mpiexec --bind-to none -mca plm isolated`
>       Is there a way to pass along command line options to how `ctest` invokes mpiexec/mpirun?

yes. Use `MPIEXEC_EXECUTABLE` for the excutable, `MPIEXEC_NUMPROC_FLAG` for -n or -np, `MPIEXEC_PREFLAGS` for `--bind-to none -mca plm isolated`, you may need  to replace spaces with semicolons

**Comment by ltalirz** [comment]
Thanks @ye-luo ! 
The  `MPIEXEC_PREFLAGS` work now but, for some reason, the max numprocs don't seem to be picked up (it's still trying to run on 4 mpi processes).
I thought I might have a typo in the name of the variable but it doesn't seem like it (?).
Do you have an idea why that might be?

**Comment by ye-luo** [comment]
actually MPIEXEC_MAX_NUMPROCS is ignored because it can be accurately used. The ctest connected tests not only include test-suite but also unit tests. Some unit tests intentionally run 9 MPI processes to test parallel eigensolver. If you would like to run all tests from test-suites with 2 MPI ranks, add `-DTESTCODE_NPROCS=2` to cmake. When you run tests, use `ctest -j N` to define N slots for running tests. If there are slots available, ctest tries to run as many tests as possible. If there is not enough slots, the next test waits for previous tests to finish and free slots. If there is no previous tests and the number of slots is still insufficient, the request test will run and oversubscribe the slots.

**Comment by ltalirz** [comment]
Thanks a lot @ye-luo for the hints!

Most tests pass now; there's one failing because of numerical result exceeding threshold (I wonder whether that really points to an error or whether instead the threshold should be adjusted)
```
pw_vdw - vdW-DF3-opt1.in: **FAILED**.
geom
    ERROR: absolute error 8.95e-03 greater than 5.00e-03. (Test: 230.28137.  Benchmark: 230.29032.)
```

It also seems that passing the additional arguments for mpiexec cause problems but only for some of the tests, see e.g. [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=356077&view=logs&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d&l=2810).
I guess the other tests that are passing also use mpiexec to  run, right?

I'll now try replacing spaces by semicolons to see whether that fixes it

**Comment by ye-luo** [comment]
0. You confirmed my previous comment "you maybe need to replace spaces with semicolons" should be "should be" ;)
1. vdW-DF3-opt1 it is a question directly to QE
2. other failing tests "Alternatively, you can use the --oversubscribe option to ignore the
number of available slots when deciding the number of processes to
launch." should make them pass.

**Comment by ltalirz** [comment]
I'm putting this PR on hold for the moment.

Besides the patches that would be needed for QE, there is the issue that libxc is needed for the tests to pass.
libxc for MacOS was only added with libxc version 5, but on OSX the build is missing the fortran bindings.

I've pinged the maintainer at https://github.com/conda-forge/libxc-feedstock/issues/15 ; once the fortran bindings for libxc are present, one can continue with this PR (or we should exclude any tests that require libxc).

**Comment by ltalirz** [comment]
hm... it turns out that the gnu make installation route fails for QE 6.8 because it tries to build some CUDA-related libraries (devxlib), which I guess are built unnecessarily (?). I don't seem to be able to convince the build not to do this https://github.com/conda-forge/qe-feedstock/pull/19

Given these headaches, and given that the tests with libxc seem to pass fine, at least on Linux, perhaps the easiest is to remove libxc here again and simply ignore the test outputs.
Let's hope that this situation will improve with the next release.

**Comment by ltalirz** [comment]
Ok, now after removing libxc something seems to have changed in the conda package ecosystem, introducing some dependency conflict (as usual, the error messages from the conda resolver are too cryptic to understand what the cause of hte issue is).

I cannot devote more time to this at the moment - if someone wants to take over with the QE 6.8 release, please go ahead.

**Comment by ltalirz** [comment]
@conda-forge-admin , please rerender

**Comment by ltalirz** [comment]
Linux: 
```
99% tests passed, 1 tests failed out of 308
The following tests FAILED:
	256 - system--pw_vdw-correctness (Failed)
```

MacOS:
```
87% tests passed, 40 tests failed out of 308

	188 - system--pw_realspace-correctness (Failed)
	193 - system--pw_relax-correctness (Failed)
	200 - system--pw_scf-correctness (Failed)
	222 - system--pw_scf--scf-ppcg-k (Failed)
	224 - system--pw_spinorbit-correctness (Failed)
	230 - system--pw_uspp-correctness (Failed)
	241 - system--pw_uspp--uspp-ppcg-k (Failed)
	247 - system--pw_vc-relax-correctness (Failed)
	256 - system--pw_vdw-correctness (Failed)
	267 - system--pw_workflow_relax_relax-correctness (Failed)
	270 - system--pw_workflow_scf_dos-correctness (Failed)
	273 - system--pw_workflow_vc-relax_dos-correctness (Failed)
	276 - system--pw_workflow_vc-relax_scf-correctness (Failed)
	280 - system--cp_h2o-correctness (Failed)
	285 - system--cp_h2o_exx-correctness (Failed)
	290 - system--cp_h2o_wf-correctness (Failed)
	293 - system--cp_o2-correctness (Failed)
	300 - system--cp_si-correctness (Failed)
	304 - system--cp_sio2-correctness (Failed)
```

@ye-luo It looks like the linux cmake build is working fine; one MacOS it looks like all of the "correctness" tests are failing 
I also see things like 
```
WARNING: cannot find program: XXXXXX/test-suite/run-pw.sh.
WARNING: cannot find program: XXXXXX/test-suite/extract-pw.sh.
```
in the MacOS build.

Any ideas what could be causing this?

**Comment by ye-luo** [comment]
Linux. Could you add  `--outpu-on-failure` to `ctest` to see what is the actual failure?
Mac. Something went wrong around https://gitlab.com/QEF/q-e/-/blob/develop/test-suite/CMakeLists.txt#L65
So `<build>/test-suite/userconfig` still carries XXXXXX from the tmp file.
Unfortunately. I don't have Mac around me. Could you help debugging?

**Comment by ltalirz** [comment]
> Linux. Could you add `--outpu-on-failure` to `ctest` to see what is the actual failure? It is like some minor things due to the numerical round error.

Thanks, I will!

> Mac. Something went wrong around https://gitlab.com/QEF/q-e/-/blob/develop/test-suite/CMakeLists.txt#L65
So <build>/test-suite/userconfig still carries XXXXXX from the tmp file.
Unfortunately. I don't have Mac around me. Could you help debugging?

Ah great, thanks for pointing to the issue.
Indeed, there is a difference between the GNU and the BSD `sed` when replacing "in-place".

If we don't mind having backup copies around, we can use the portable `sed -i.bak`, see
https://stackoverflow.com/a/22084103/1069467

Can you prepare the patch?

**Comment by ltalirz** [comment]
Sorry, just realized that `--output-on-failure` was already enabled.
It is indeed a precision issue

```
266/308 Test #256: system--pw_vdw-correctness ........................................***Failed    1.00 sec
Test id: 010121.
Benchmark: git.

pw_vdw - beef-spin.in: Passed.

pw_vdw - beef.in: Passed.

pw_vdw - rVV10.in: Passed.

pw_vdw - vdW-DF3-opt1.in: **FAILED**.
geom
    ERROR: absolute error 8.95e-03 greater than 5.00e-03. (Test: 230.28137.  Benchmark: 230.29032.)

pw_vdw - vdW-DF3-opt2.in: Passed.

pw_vdw - vdw-d2.in: Passed.

pw_vdw - vdw-d3.in: Passed.

pw_vdw - vdw-mbd.in: Passed.

pw_vdw - vdw-ts.in: Passed.

pw_vdw - xdm.in: Passed.

All done. ERROR: only 9 out of 10 tests passed.
Failed test in:
	$SRC_DIR/build/test-suite/pw_vdw/

```

**Comment by ye-luo** [comment]
See https://gitlab.com/QEF/q-e/-/merge_requests/1690

**Comment by ltalirz** [comment]
Linux build: unchanged (one should probably adjust the precision in the test).
```
99% tests passed, 1 tests failed out of 308
The following tests FAILED:
	256 - system--pw_vdw-correctness (Failed)
```

MacOS build: much better than before:
```
98% tests passed, 6 tests failed out of 308

The following tests FAILED:
	 23 - test_qe_lapack_zdotc (Failed)
	148 - system--pw_metaGGA-correctness (Failed)
	200 - system--pw_scf-correctness (Failed)
	222 - system--pw_scf--scf-ppcg-k (Failed)
	230 - system--pw_uspp-correctness (Failed)
	241 - system--pw_uspp--uspp-ppcg-k (Failed)
```

Examples of errors found (see also full log at [macos_tests.log](https://github.com/conda-forge/qe-feedstock/files/7774740/macos_tests.log)).

```

241/308 Test #241: system--pw_uspp--uspp-ppcg-k ......................................***Failed  Required regular expression not found. Regex=[JOB DONE
libxc needed for this functional
]  2.52 sec


pw_uspp - uspp-ppcg-k.in: **FAILED**.
Different sets of data extracted from benchmark and test.
    Data only in benchmark: f1, eh1, e1, n1.
```

These failing tests should probably be investigated (e.g. it is not clear to me why a test requiring libxc is even executed), but at this point I'm going to go ahead and merge this to have a v6.8 build.

Further improvements can be made in a build number 1.

**Comment by ltalirz** [comment]
Thanks a lot for your help @ye-luo - it's finally time to move the conda-forge QE feedstock to the cmake route
